Add Google Cloud Storage as backend

diptadas opened the pull request:
### Environment variables:

```

GOOGLE_PROJECT_ID=gcp-project-id
GOOGLE_APPLICATION_CREDENTIALS=path-to-json-file

```

### Environment variables for test:

```

RESTIC_TEST_GS_PROJECT_ID=gcp-project-id
RESTIC_TEST_GS_APPLICATION_CREDENTIALS=path-to-json-file
RESTIC_TEST_GS_REPOSITORY=gs:us-central1/test-bucket

```

### Init repository in default location:

```

$ restic -r gs:/bucket-name init
or
$ restic -r gs:///bucket-name init

```

### Init repository in specific location:

```

$ restic -r gs:us-central1/bucket-name init
or
$ restic -r gs://us-central1/bucket-name init

```

TamalSaha commented:
@fd0 , you will need a paid Google cloud account and issue a service account json file from that to run these tests. If you would like us to provide the account, please share a gmail account. I can give you access to our test Google cloud project that you will be able to use to run tests.

fd0 commented:
Thank you very much for your contribution, I'll have a look as soon as I get some spare time!

fd0 commented on src/restic/backend/gs/config.go:
```diff
@@ -0,0 +1,71 @@
+	}
+	// use the first entry of the path as the endpoint and the
+	// remainder as bucket name and prefix
+	path := strings.SplitN(s, "/", 3)
```
Is the endpoint relevant for other commands besides `init`?

I'd suggest using a format of `gcs:ENDPOINT:PATH`, so that it is consistent with the `s3` backend.

fd0 commented on src/restic/backend/gs/config.go:
```diff
@@ -0,0 +1,71 @@
+	options.Register("gs", Config{})
+}
+
+const defaultPrefix = "restic"
```
Drop the default prefix, that's also a deprecated setting of the `s3` backend. If no path is given in the location, assume it is `/`.

fd0 commented on src/restic/backend/gs/config_test.go:
```diff
@@ -0,0 +1,121 @@
+	s   string
+	cfg Config
+}{
+	{"gs://us-central1/bucketname", Config{
```
Same here, drop the double slash, that's deprecated syntax :)

fd0 commented on src/restic/backend/gs/gs.go:
```diff
@@ -0,0 +1,461 @@
+// make sure that *Backend implements backend.Backend
+var _ restic.Backend = &Backend{}
+
+const defaultLayout = "default"
```
This line is not needed, you're already using the `DefaultLayout`.

fd0 commented on src/restic/backend/gs/gs.go:
```diff
@@ -0,0 +1,461 @@
+
+	raw, err := ioutil.ReadFile(jsonKeyPath)
+	if err != nil {
+		return nil, err
```
Use `errors.Wrap(err, "ReadFile")` (which uses https://github.com/pkg/errors) here to wrap the error with a stack trace. The same goes for all other errors coming from outside restic.

fd0 commented on src/restic/backend/gs/gs.go:
```diff
@@ -0,0 +1,461 @@
+	}
+
+	// Create bucket if not exists
+	if _, err := service.Buckets.Get(be.bucketName).Do(); err != nil {
```
Split `Open` into two different functions `Open` (which assumes the bucket is there) and `Create` (which tests for and creates the bucket if necessary). For an example see the `s3` backend.

fd0 commented on src/restic/backend/gs/gs.go:
```diff
@@ -0,0 +1,461 @@
+// IsNotExist returns true if the error is caused by a not existing file.
+func (be *Backend) IsNotExist(err error) bool {
+	debug.Log("IsNotExist(%T, %#v)", err, err)
+	return os.IsNotExist(err)
```
Is that enough? For the s3 backend, we've had to test for more things: https://github.com/restic/restic/blob/master/src/restic/backend/s3/s3.go#L103

fd0 commented on src/restic/backend/gs/gs.go:
```diff
@@ -0,0 +1,461 @@
+	info, err := be.service.Objects.Insert(be.bucketName,
+		&storage.Object{
+			Name: objName,
+			Size: uint64(size),
```
Is it necessary to know the size of the data beforehand? Otherwise we could drop the `getRemainingSize()`, which also has been removed from the s3 backend...

fd0 commented on src/restic/backend/gs/gs.go:
```diff
@@ -0,0 +1,461 @@
+
+	// wrap the reader so that net/http client cannot close the reader, return
+	// the token instead.
+	rd = preventCloser{
```
Is this needed for the gcs client? Does it really close the reader? You can run the integration tests, which checks that. Otherwise drop the preventCloser{}.

fd0 reviewed (changes requested):
First round of code review is done, please let me know what you think. Thanks for giving this a try!

fd0 commented on src/restic/backend/gs/gs.go:
```diff
@@ -0,0 +1,461 @@
+func (be *Backend) Close() error { return nil }
+
+// Rename moves a file based on the new layout l.
+func (be *Backend) Rename(h restic.Handle, l backend.Layout) error {
```
Remove the Rename() function, it was used to implement the migrate command for the s3 backend, it isn't used for GCS.

tamalsaha commented on src/cmds/restic/global.go:
```diff
@@ -358,6 +359,31 @@ func parseConfig(loc location.Location, opts options.Options) (interface{}, erro
 		debug.Log("opening s3 repository at %#v", cfg)
 		return cfg, nil
 
+	case "gs":
```
GCS's cli gsutil uses `gs://` prefix https://cloud.google.com/storage/docs/gsutil

fd0 replied:
Ok, then we'll use that.

tamalsaha commented on src/restic/backend/gs/config.go:
```diff
@@ -0,0 +1,71 @@
+// Config contains all configuration necessary to connect to an gcs compatible
+// server.
+type Config struct {
+	Location    string
```
I think Location should be removed from here. According to api,
```
	// Location: The location of the bucket. Object data for objects in the
	// bucket resides in physical storage within this region. Defaults to
	// US. See the developer's guide for the authoritative list.
	Location string `json:"location,omitempty"`
```

So, if user wants to use a different location, they can create they bucket ahead of time. This will be similar to how S3 behaves.

diptadas commented:
**Specific location:**
```$ restic -r gs:location:bucketname/prefix/directory init```

**Default location:**
```$ restic -r gs:bucketname/prefix/directory init```

tamalsaha commented:
@diptadas , we should change the URL to `gs:bucket:/[prefix]` .  There should be a `:` after bucket name.

diptadas commented:
**Format:** `gs:bucket:/[prefix] `

diptadas committed:
Add Google Cloud Storage as backend

Environment variables:
GOOGLE_PROJECT_ID=gcp-project-id
GOOGLE_APPLICATION_CREDENTIALS=path-to-json-file

Environment variables for test:
RESTIC_TEST_GS_PROJECT_ID=gcp-project-id
RESTIC_TEST_GS_APPLICATION_CREDENTIALS=path-to-json-file
RESTIC_TEST_GS_REPOSITORY=gs:us-central1/test-bucket

Init repository:
$ restic -r gs:bucket:/[prefix] init

fd0 commented:
I'll take over the remaining work, successor is #1134. Thank you!
